### ISSUE TITLE
Sharing StyleLayers between WorkerTiles is dangerous

### ISSUE THREAD

**Issue by jfirebaugh** [mapbox/mapbox-gl-js#3479]
`StyleLayer`s are shared between `WorkerTile`s, via the worker `StyleLayerIndex`. But `StyleLayer`s contain state consisting of the evaluated result of style properties at a certain zoom level, and `WorkerTile`s may have different zoom levels.

Currently, we avoid any issues that this shared state could cause by carefully calling `StyleLayer#recalculate` at the entry point to every async callback in `WorkerTile`. But this is relatively fragile and error prone, and also requires reevaluation of previously-evaluated properties.

A better solution would be to have each `WorkerTile` call `StyleLayer#recalculate` once for each necessary layer, and store the logical result state independently of any other `WorkerTile`s. This would likely require a significant refactor of the internals of style recalculation (refs #2739, #3044).

**Comment by DannyDelott** [comment]
This is interesting.

I'm noticing a bug where `Map#queryRenderedFeatures` throws an `AssertionError: false === true` at lower zoom levels on my custom layer. When I zoom in closer on the map and try again, it is able to get the features from my custom layer no problem. It's hard to reproduce, so I haven't filed a bug for it-- but this might be related?

What areas of the library are affected by this issue?

**Comment by jfirebaugh** [comment]
I don't believe this is causing any bugs currently -- this is a refactoring issue noting how we could improve the code in the future. An assertion error in `queryRenderedFeatures` sounds like a separate issue.

**Comment by kristfal** [comment]
@DannyDelott just encountered the same issue while running `V0.23.0`. The assertion error should be fixed by #3233.
